src/twisted/names/client.py

```diff
@@ -217,13 +217,26 @@ def _connectedProtocol(self, interface=''):
         Return a new L{DNSDatagramProtocol} bound to a randomly selected port
         number.
         """
+        failures = 0
         proto = dns.DNSDatagramProtocol(self, reactor=self._reactor)
+
         while True:
             try:
                 self._reactor.listenUDP(dns.randomSource(), proto,
                                         interface=interface)
-            except error.CannotListenError:
-                pass
+            except error.CannotListenError as e:
+                failures += 1
+
+                if (hasattr(e.socketError, "errno") and
```

Comment:
I appreciate that retrying in a hot loop is unlikely to fix `EMFILE`, but this could be made much more reliable in FD-pressure circumstances by doing a *slow* retry with a `LoopingCall` and not violating the API contract of Resolver.queryUDP (i.e.: synchronously raising where it doesn't document exceptions and specifically says that it returns a `Deferred` that fails with `TimeoutError`.)